pkg/spoke/statusfeedback/reader.go

```diff
@@ -0,0 +1,121 @@
+package statusfeedback
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/util/jsonpath"
+	workapiv1 "open-cluster-management.io/api/work/v1"
+	"open-cluster-management.io/work/pkg/spoke/statusfeedback/rules"
+)
+
+type StatusReader struct {
+	wellKnownStatus rules.WellKnownStatusRuleResolver
+}
+
+func NewStatusReader() *StatusReader {
+	return &StatusReader{
+		wellKnownStatus: rules.DefaultWellKnownStatusRule(),
+	}
+}
+
+func (s *StatusReader) GetValuesByRule(obj *unstructured.Unstructured, rule workapiv1.FeedbackRule) ([]workapiv1.FeedbackValue, error) {
+	errs := []error{}
+	values := []workapiv1.FeedbackValue{}
+
+	switch rule.Type {
+	case workapiv1.WellKnownStatusType:
+		gvk := obj.GroupVersionKind()
+		paths := s.wellKnownStatus.GetPathsByKind(gvk.GroupKind())
+		if len(paths) == 0 {
+			break
+		}
+
+		for _, path := range paths {
+			value, err := getValueByJsonPath(path.Name, path.Path, obj)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			if value == nil {
+				continue
+			}
+			values = append(values, *value)
+		}
+	case workapiv1.JSONPathsType:
+		for _, path := range rule.JsonPaths {
+			// skip if version is specified and the object version does not match
+			if len(path.Version) != 0 && obj.GroupVersionKind().Version != path.Version {
+				continue
+			}
+
+			value, err := getValueByJsonPath(path.Name, path.Path, obj)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			if value == nil {
+				continue
+			}
+			values = append(values, *value)
+		}
+	}
+
+	return values, utilerrors.NewAggregate(errs)
+}
+
+func getValueByJsonPath(name, path string, obj *unstructured.Unstructured) (*workapiv1.FeedbackValue, error) {
+	j := jsonpath.New(name).AllowMissingKeys(true)
+	err := j.Parse(fmt.Sprintf("{.status%s}", path))
+	if err != nil {
+		return nil, err
+	}
+
+	results, err := j.FindResults(obj.UnstructuredContent())
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(results) == 0 || len(results[0]) == 0 {
+		// no results are found here.
+		return nil, nil
+	}
+
+	// as we only support simple JSON path, we can assume to have only one result (or none, filtered out above)
+	value := results[0][0].Interface()
+	var fieldValue workapiv1.FieldValue
+	switch t := value.(type) {
+	case int64:
```

Comment:
the int64 is converted to 32 implicitly here. i remember it's due to the limitation of unstructured json codec, it's taking all the numeric values to int64, is it? fwiw i think we should assert the overflowing here defensively here such as `int64(int32(t)) == int64(t)`?

Reply:
yes, that is the limitation on json codec. If we assert overflowing here, that will cause that an int64 value cannot be returned, like observedGeneration. I am think we should define this as int64 in API. WDYT?